### PR TITLE
Reimplement Zelda HLE

### PR DESCRIPTION
**Pull request by delroth**
It is finally there. Now that all the priority high/critical bugs have been fixed, I think this is ready for review.

Here are the known remaining bugs:

* [Pikmin 2 save sound too high pitched](https://code.google.com/p/dolphin-emu/issues/detail?id=8855&q=nzhle) — already a bug in current implementation (and minor).
* [No echo in MK:DD](https://code.google.com/p/dolphin-emu/issues/detail?id=8034&q=nzhle) — already a bug in current implementation.
* [NTSC IPL sounds missing](https://code.google.com/p/dolphin-emu/issues/detail?id=7981&q=nzhle) — IPL is completely broken with the current HLE implementation (crashing, I think?) so this is ultimately better.
* [IPL sounds don't end properly](https://code.google.com/p/dolphin-emu/issues/detail?id=7980&q=nzhle) — same thing, overall better than previous implementation.

There are two remaining bugs that only appear in "strict" mode, which don't seem to have noticeable impact on the games.

I'll let @JMC47 describe how much NZHLE fixes and improves. I don't play games so it's hard to say :)

### CODE REVIEW

File: Source/Core/Core/HW/DSPHLE/MailHandler.cpp [hunk 1]
```diff
@@ -14,9 +15,20 @@ CMailHandler::~CMailHandler()
 	Clear();
 }
 
-void CMailHandler::PushMail(u32 _Mail)
+void CMailHandler::PushMail(u32 _Mail, bool interrupt)
 {
-	m_Mails.push(_Mail);
+	if (interrupt)
+	{
+		if (m_Mails.size() == 0)
```
Comment (lioncash) [inline Source/Core/Core/HW/DSPHLE/MailHandler.cpp]: `if (m_Mails.empty())`

File: Source/Core/Core/HW/DSPHLE/MailHandler.cpp [hunk 2]
```diff
@@ -14,9 +15,20 @@ CMailHandler::~CMailHandler()
...
+		{
+			DSP::GenerateDSPInterruptFromDSPEmu(DSP::INT_DSP);
+		}
+		else
+		{
+			m_Mails.front().second = true;
+		}
+	}
+	m_Mails.push(std::make_pair(_Mail, false));
```
Comment (lioncash) [inline Source/Core/Core/HW/DSPHLE/MailHandler.cpp]: Can in-place insert with

`m_Mails.emplace(_Mail, false);`

File: Source/Core/Core/HW/DSPHLE/UCodes/UCodes.cpp [hunk 3]
```diff
@@ -51,27 +52,18 @@ UCodeInterface* UCodeFactory(u32 crc, DSPHLE* dsphle, bool wii)
...
-	case 0x6c3f6f94: // Zelda TP - PAL
-	case 0xd643001f: // Mario Galaxy - PAL / Wii DK Jungle Beat - PAL
-		INFO_LOG(DSPHLE, "CRC %08x: Zelda Wii ucode chosen\n", crc);
+	case 0x6ca33a6d: // Zelda TP GC - US
+	case 0xd643001f: // Super Mario Galaxy - US
+	case 0x6ba3b3ea: // GC IPL - PAL
+	case 0x24b22038: // GC IPL - US
+	case 0x2fcdf1ec: // Zelda FSA - US
+	case 0x4be6a5cb: // Pimin 1 GC - US
+	case 0x267fd05a: // Pimin 1 GC - PAL
```
Comment (JosJuice) [inline Source/Core/Core/HW/DSPHLE/UCodes/UCodes.cpp]: Typo?

File: Source/Core/Core/HW/DSPHLE/MailHandler.cpp [hunk 4]
```diff
@@ -95,9 +118,10 @@ void CMailHandler::DoState(PointerWrap &p)
 		// Restore queue.
 		for (int i = 0; i < sz; i++)
 		{
-			u32 value = temp.front();
+			u32 value = temp.front().first;
+			bool interrupt = temp.front().second;
 			temp.pop();
-			m_Mails.push(value);
+			m_Mails.push(std::make_pair(value, interrupt));
```
Comment (lioncash) [inline Source/Core/Core/HW/DSPHLE/MailHandler.cpp]: emplace can also be used here and with the other two queue pushes above this in the diff.

**Comment by Tilka** [comment]
Yup, lgtm.

**Comment by BhaaLseN** [comment]
LGTM code-wise (not tested)

**Comment by lioncash** [comment]
Looks good style-wise.

**Comment by wasteofoxygen** [comment]
hnnnnnnng

File: Source/Core/Core/HW/DSPHLE/UCodes/UCodes.cpp [hunk 5]
```diff
@@ -51,27 +52,18 @@ UCodeInterface* UCodeFactory(u32 crc, DSPHLE* dsphle, bool wii)
...
+	case 0xd643001f: // Super Mario Galaxy - US
+	case 0x6ba3b3ea: // GC IPL - PAL
+	case 0x24b22038: // GC IPL - US
+	case 0x2fcdf1ec: // Zelda FSA - US
+	case 0x4be6a5cb: // Pikmin 1 GC - US
+	case 0x267fd05a: // Pikmin 1 GC - PAL
+	case 0x42f64ac4: // Luigi's Mansion - US
+	case 0x56d36052: // Super Mario Sunshine - US
+	case 0x6c3f6f94: // Zelda TP Wii - US
+	case 0xeaeb38cc: // Pikmin 1/2 New Play Control Wii - US
```
Comment (magumagu) [inline Source/Core/Core/HW/DSPHLE/UCodes/UCodes.cpp]: It looks like the 0xb7eb9a9c CRC disappeared at some point; is that intentional?

Comment (delroth) [inline Source/Core/Core/HW/DSPHLE/UCodes/UCodes.cpp]: Nope, looks like nobody ever tested that game with nzhle. I'll gladly take PRs, I'm currently stuck in a tent without power or WiFi :)

Comment (BhaaLseN) [inline Source/Core/Core/HW/DSPHLE/UCodes/UCodes.cpp]: A PR isn't as easy as you'd think...as it is now, it won't run with or without the CRC added for ZeldaUCode, nor with various flags (it gets HALTed with default protocol, and doesn't boot/gets stuck on light protocol).
Pikmin Wii PAL (R9IP01) that is.

**Pull request merged**